core/src/main/java/org/geysermc/geyser/command/GeyserCommand.java

```diff
@@ -25,74 +25,194 @@
+     */
+    public static final CommandMeta.Key<Boolean> PLAYER_ONLY = CommandMeta.Key.of(Boolean.class, "player-only", meta -> false);
+
+    /**
+     * The second literal of the command. Note: the first literal is {@link #rootCommand()}.
+     */
+    @NonNull
+    private final String name;
 
-    protected final String name;
     /**
      * The description of the command - will attempt to be translated.
      */
-    protected final String description;
-    protected final String permission;
+    @NonNull
+    private final String description;
 
-    private List<String> aliases = Collections.emptyList();
+    /**
+     * The permission node required to run the command, or blank if not required.
+     */
+    @NonNull
+    private final String permission;
 
-    public abstract void execute(@Nullable GeyserSession session, GeyserCommandSource sender, String[] args);
+    /**
+     * The default value of the permission node.
+     * A null value indicates that the permission node should not be registered whatsoever.
+     */
+    @Nullable
+    private final TriState permissionDefault;
 
     /**
-     * If false, hides the command from being shown on the Geyser Standalone GUI.
-     *
-     * @return true if the command can be run on the server console
+     * True if the command can be executed by non players
      */
-    @Override
-    public boolean isExecutableOnConsole() {
-        return true;
-    }
+    private final boolean executableOnConsole;
+
+    /**
+     * True if the command can only be run by bedrock players
+     */
+    private final boolean bedrockOnly;
 
     /**
-     * Used in the GUI to know what subcommands can be run
-     *
-     * @return a list of all possible subcommands, or empty if none.
+     * The aliases of the command {@link #name}
      */
+    protected List<String> aliases = Collections.emptyList();
+
+    public GeyserCommand(@NonNull String name, @Nullable String description,
+                         @Nullable String permission, @Nullable TriState permissionDefault,
+                         boolean executableOnConsole, boolean bedrockOnly) {
+
+        if (name == null || name.isBlank()) {
+            throw new IllegalArgumentException("Command cannot be null or blank!");
+        }
+
+        this.name = name;
+        this.description = description != null ? description : "";
+        this.permission = permission != null ? permission : "";
+        this.permissionDefault = permissionDefault;
+
+        if (bedrockOnly && executableOnConsole) {
+            throw new IllegalArgumentException("Command cannot be both bedrockOnly and executableOnConsole");
+        }
+
+        this.executableOnConsole = executableOnConsole;
+        this.bedrockOnly = bedrockOnly;
+    }
+
+    public GeyserCommand(@NonNull String name, @NonNull String description, @Nullable String permission, @NonNull TriState permissionDefault) {
```

Comment:
Why is `permissionDefault` marked `@NonNull` here, but `@Nullable` in the other constructor above?

Reply:
forgot to change both